### PR TITLE
Fix RwLock::try_write

### PR DESCRIPTION
**Pull request by jgallagher**
Previously, `try_write` actually only obtained shared read access (but would return a `RwLockWriteGuard` if that access was successful).

Also updates the docs for `try_read` and `try_write`, which were leftover from when those methods returned `Option` instead of `Result`.

### CODE REVIEW

**Comment by brson** [comment]
Nominating for beta by request of the author.

File: src/libstd/sync/rwlock.rs [hunk 1]
```diff
@@ -171,14 +171,16 @@ impl<T: ?Sized> RwLock<T> {
         RwLockReadGuard::new(&*self.inner, &self.data)
     }
 
-    /// Attempts to acquire this lock with shared read access.
+    /// Attempts to acquire this rwlock with shared read access.
+    ///
+    /// If the access could not be granted at this time, then `Err` is returned.
+    /// Otherwise, an RAII guard is returned which will release the shared access
+    /// when it is dropped.
```
Comment (alexcrichton) [inline src/libstd/sync/rwlock.rs]: I believe that a poisoned rwlock will also return an `Err` even if it was successful in the `try_read`, so technically the lock could be locked if `Err` is returned. Perhaps the docs could be updated slightly to account for this?

**Comment by alexcrichton** [comment]
Nice catch @jgallagher! Just one small nit from me on the doc updates and otherwise this is good to go.

File: src/libstd/sync/rwlock.rs (hunk 1 above)
Comment (jgallagher) [inline src/libstd/sync/rwlock.rs]: I think the pre-existing `Failure` section covers this, but it's outside the diff of this PR (https://github.com/jgallagher/rust/blob/rwlock-try-write/src/libstd/sync/rwlock.rs#L187-L190). Is what it states incorrect, or should this paragraph be reworded to include it somehow?

Comment (alexcrichton) [inline src/libstd/sync/rwlock.rs]: Hm ok so upon re-reading my original though that this indicates `Err` means the lock wasn't grabbed isn't quite right, it just means that if it couldn't be grabbed `Err` is returned, so looks good to me!

**Pull request merged**